File: PascalTri.py
"""
// Time Complexity : mentioned along code
// Space Complexity : mentioned along code
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach
"""

class Solution:
    """
    RECURSIVE
    Time complexity : o(n^2)
    Space complexity : o(k)  
    """
    """

    def helper(self, prev, cur_ind, k):
        
        
        cur_row = [0 for _ in range(cur_ind)] #initializing current row
        
        cur_row[0] = 1 #placing 1's at 0th and last index of the row
        cur_row[-1] = 1
        
        for j in range(1, len(cur_row)-1): #filling up the remaining cells
            cur_row[j] = prev[j-1] + prev[j]
        
        #prev = cur_row
        #cur_ind +=1
        if cur_ind == k+1: base case
            return cur_row
         
        return self.helper( cur_row, cur_ind+1, k)
        """
        
    
    def getRow(self, rowIndex: int) -> List[int]:
        """
        ITERATIVE
        Time complexity : o(n^2), n -> number of rows
        Space complexity : o(k)
        """

        prev = [] #keeps track of previous row
        
        for i in range(rowIndex+ 1):
            cur = [0 for _ in range (i+1)] #initializing the current row 
            
            cur[0],cur[-1] = 1,1 #placing 1's at 0th and last index of the row
            
            for k in range(1,len(cur)-1): #filling up the remaining cells
                
                cur[k] = prev[k-1] + prev[k]
                
            prev = cur   #updating prev
        return cur
        """
        return self.helper([1],1, rowIndex)
        """

File: pascalTri.cpp
// Time Complexity : O(N^2)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
class Solution {
public:
    vector<vector<int>> generate(int numRows) {   //numRows=5
        vector<vector<int>> answer;
        if (numRows==0)
            return answer;
        
        vector<int> temp(1,1); //first row    1
        answer.push_back(temp);
         
        if (numRows==1) //sanity check if numRows is 1
            return answer;
        
        vector<int> temp2(2,1); //second row    //1
                                                //1,1
        answer.push_back(temp2);                //1,2,1
        
         if (numRows==2) //sanity check if numRows is 2
            return answer;
        
        for (int i=2;i<numRows;i++) {
            vector<int> temp3;
            temp3.push_back(1);   // 1
            for (int j=1 ; j<i ;j++) {
                temp3.push_back(answer[i-1][j]+answer[i-1][j-1]); //1,2
            }
            temp3.push_back(1); //1,2,1
            answer.push_back(temp3); 
        }
        
        return answer;
    }
};